invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,210 @@
+        self._database = database
+        self._migrations: set[Migration] = set()
+
+    def register_migration(self, migration: Migration) -> None:
+        """Registers a migration."""
+        if not isinstance(migration.db_version, int) or migration.db_version < 1:
+            raise MigrationVersionError(f"Invalid migration version {migration.db_version}")
+        if any(m.db_version == migration.db_version for m in self._migrations):
+            raise MigrationVersionError(f"Migration version {migration.db_version} already registered")
+        self._migrations.add(migration)
+        self._logger.debug(f"Registered migration {migration.db_version}")
+
+    def run_migrations(self) -> None:
+        """Migrates the database to the latest version."""
+        with self._lock:
+            self._create_version_table()
+            sorted_migrations = sorted(self._migrations, key=lambda m: m.db_version)
+            current_version = self._get_current_version()
+
+            if len(sorted_migrations) == 0:
+                self._logger.debug("No migrations registered")
+                return
+
+            latest_version = sorted_migrations[-1].db_version
+            if current_version == latest_version:
+                self._logger.debug("Database is up to date, no migrations to run")
+                return
+
+            if current_version > latest_version:
+                raise MigrationError(
+                    f"Database version {current_version} is greater than the latest migration version {latest_version}"
+                )
+
+            self._logger.info("Database update needed")
+
+            # Only make a backup if using a file database (not memory)
+            backup_path: Optional[Path] = None
+            if isinstance(self._database, Path):
+                backup_path = self._backup_db(self._database)
+            else:
+                self._logger.info("Using in-memory database, skipping backup")
+
+            for migration in sorted_migrations:
+                try:
+                    self._run_migration(migration)
+                except MigrationError:
+                    if backup_path is not None:
+                        self._logger.error(f" Restoring from {backup_path}")
+                        self._restore_db(backup_path)
+                    raise
+            self._logger.info("Database updated successfully")
+
+    def _run_migration(self, migration: Migration) -> None:
+        """Runs a single migration."""
+        with self._lock:
+            current_version = self._get_current_version()
+            try:
+                if current_version >= migration.db_version:
+                    return
+                migration.migrate(self._cursor)
+                # Migration callbacks only get a cursor; they cannot commit the transaction.
+                self._conn.commit()
+                self._set_version(db_version=migration.db_version, app_version=migration.app_version)
```

Comment:
In the unlikely but conceivable case that there is a crash/power outage between the `commit()` in the line above and the `_set_version()` here you could end up with a database with the newer version's tables but the older  version number.  When the app is restarted the migration system will re-run the migration and then fail because the new tables and indexes are already there. I think `_set_version()` should be moved above the `commit()` to avoid this. The call to `commit()` in `_set_version()` should then be removed.